text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,559 @@
+programming, adopted to the needs of distributed systems.
+
+OpenTelemetry is separated into two layers: **aspects** which crosscut and 
+intertwine with a program's functional concerns, and cannot be encapsulated. In 
+this architecture, each cross-cutting concern is modeled as an independent 
+subsystem. Multiple aspects - including the tracing and baggage systems provided 
+by OpenTelemetry - then share the same underlying **context propagation** 
+system, which allows these cross-cutting concerns to store and access their 
+data across the lifespan of a distributed transaction.
+
+# Aspects
+
+## Observability API
+Distributed tracing is one example of an aspect. Tracing code is interleaved 
+with regular code, and ties together indepentent code modules which would 
+otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+A second observability aspect is correlations. Correlations are labels for 
+metrics, where the value of the label may be defined anywhere in the 
+transaction. Correlations are like Baggage, but are write only - the values can 
+only be used for observability.
+
+The observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+## Baggage API
+
+In addition to observability, OpenTelemetry provides a simple mechanism for 
+propagating arbitrary user data, called Baggage. This mechanism is not related
+to tracing or observability, but it uses the same context propagation layer. 
+
+Baggage may be used to model new aspects which would benefit from the same 
+transaction-level context as tracing, e.g., for identity, versioning, and 
+network switching. 
+
+To manage the state of these cross-cutting concerns, the Baggage API provides a 
+set of functions which read, write, and propagate data.
+
+**SetBaggage(context, key, value) -> context**  
+To record the distributed state of an aspect, the Baggage API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+**GetBaggage(context, key) -> value**  
+To access the distributed state of an aspect, the Baggage API provides a 
+function which takes a context and a key as input, and returns a value.
+
+**RemoveBaggage(context, key) -> context**  
+To delete distributed state from an aspect, the Baggage API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**ClearBaggage(context) -> context**  
+To avoid sending baggage to an untrusted downstream process, the Baggage API 
+provides a function remove all baggage from a context.
+
+**GetBaggageHTTPPropagator() -> (HTTPExtract, HTTPInject)**  
```

Comment:
I think that's the intention, at least if it's not I'm +1 for making that the case. I think the idea was to enable people to implement baggage however they wanted to, and have in the API a line of code that bootstraps whatever BaggageImpl was set.

That said, the separate of API and SDK doesn't always make sense to me. I would think we just have a propagators object with an implicit global, and add (and possibly remove) propagators from there:

```
# as the sdk is instantiated, handle adding it's own extractor / injectors

PROPAGATOR_GLOBAL.extractors.add(BaggageExtractor())
PROPAGATOR_GLOBAL.injectors.add(BaggageInjector())
```

Reply:
@yurishkuro what @toumorokoshi says is correct: the intention here is that there should be a global propagator, which each aspect can chain it's individual propagator to. But since each aspect is completely separate, there has to be an interface for them to write against.

It also may be the case that a concern does not want to use the global propagator, because they have to do additional work at the injection and extraction call site. So having Propagators available on their own, and not requiring them to use the global, is also important.

Reply:
fine, feel free to resolve.